### ISSUE TITLE
`F.interpolate` returns unexpected result when dealing with `2x2` image

### ISSUE THREAD

**Issue by ZisIsNotZis** [pytorch/pytorch#30565]
## 🐛 Bug

Generally, when shrinking a image to half of its height/weight, `F.interpolate` with `mode='bilinear'` and `align_corners=False` should be equivalent to `F.avg_pool2d(2)`, and I use this fact to try out things like "will using better interpolation rather than pooling get better result" and etc.

But I found my model will fail to optimize if I change `F.interpolate` to `F.avg_pool2d` (e.g. having 10% accuracy where there are 10 classes). After a lot of separation of modules, I finally found the problem above.

It seems that `F.interpolate(a, 1, ...)` with `2x2 a` actually returns `a[...,:1,:1]` currently. I believe this should be a bug.

## To Reproduce

```python
a = torch.rand(1, 1, 2, 2)
b = F.interpolate(a, 1, mode='bilinear')

b==a, b==a.mean((2,3),True)
```

outputs

```python
(tensor([[[[ True, False],
           [False, False]]]]), tensor([[[[False]]]]))
```

## Expected behavior

A `[[False, False], [False, False]]` Tensor (generally) and a `True` tensor

## Environment

```
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 19.10
GCC version: (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 440.26
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.2.0
[pip] torchvision==0.4.0a0
[conda] _pytorch_select           0.2                       gpu_0    defaults
[conda] blas                      1.0                         mkl    defaults
[conda] mkl                       2019.4                      243    defaults
[conda] mkl-service               2.3.0            py37he904b0f_0    defaults
[conda] mkl_fft                   1.0.15           py37ha843d7b_0    defaults
[conda] mkl_random                1.1.0            py37hd6b4f25_0    defaults
[conda] pytorch                   1.2.0           cuda100py37h938c94c_0    defaults
[conda] torchvision               0.4.0           cuda100py37hecfc37a_0    defaults
```

**Comment by ZisIsNotZis** [comment]
It seems that this problem also occurs with `mode='bicubic'`, and will happen as long as the output size is `1` (regardless of the input size).

This can be easily shown by inspecting these "unbalanced“ gradient

```python
a = torch.zeros(1,1,5,5, device='cuda', dtype=torch.half, requires_grad=True)
bilinear

F.interpolate(a, 1, mode='bilinear', align_corners=False).sum().backward()
a.grad
# tensor([[[[1., 0., 0., 0., 0.],
#           [0., 0., 0., 0., 0.],
#           [0., 0., 0., 0., 0.],
#           [0., 0., 0., 0., 0.],
#           [0., 0., 0., 0., 0.]]]], device='cuda:0', dtype=torch.float16)

F.interpolate(a, 1, mode='bicubic', align_corners=False).sum().backward()
a.grad
# tensor([[[[ 2.1953, -0.1025,  0.0000,  0.0000,  0.0000],
#           [-0.1025,  0.0088,  0.0000,  0.0000,  0.0000],
#           [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
#           [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
#           [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000]]]], device='cuda:0',
#        dtype=torch.float16)

**Comment by mruberry** [comment]
Thanks for reporting this issue. 

Can you elaborate on what you'd expect from the gradients in these cases with some small examples?
Is there an interpolation mode that functions as you expect it to?

**Comment by ZisIsNotZis** [comment]
For `bilinear` with `1x1` output, it is currently returning value of the upper left corner. The gradient `[[1,0],[0,0]]` and `[[1,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0]]` (above) also indicates that it is using the upper left corner 100%.

But according to definition of bilinear, the gradient should be `[[0.25,0.25],[0.25,0.25]]` for `2x2` input, and probably `[[0,0,0,0,0],[0,0,0,0,0],[0,0,1,0,0],[0,0,0,0,0],[0,0,0,0,0]]` for `5x5` input.

`bicubic` is more complicated. But it should be some kernel matrix around the center (i.e. `(2,2)` for `5x5`) instead of around the upper left corner `(0,0)`.

For `2x2` to `1x1`, `bilinear` can be replaced by `area` or `2x2` pooing, but since the current implementation of interpolate is sort of wrong here, I believe this should be fixed in upstream here rather than making a `if-else` switch in the module code.

**Comment by mruberry** [comment]
@ZisIsNotZis thank you for elaborating!

**Comment by mruberry** [comment]
@albanD

**Comment by albanD** [comment]
The documentation about `align_corners` is so cryptic I'm not sure if this is the intended behavior or not... 
It seems like this flag was built to be used with `scale_factor` and might not work as expected with fixed sizes.
cc @SsnL  who worked on this ?

In any case, the gradients do match the forward.

**Comment by SsnL** [comment]
@albanD I didn't read this issue, but here is an attempt to explain `align_corners`. 

I'll use 1D interpolating from size `iW` to `oW` as example, but it can be easily extend to general dimensions and sizes.

When interpolating, we should think of pixels as the value of the *square*. We can two reasonable choices to map output spatial location to input spatial location. 

1. Map `0: out => 0: in`, and `oW-1: out => iW-1: in`, i.e., align the center of the corner pixels: `output[x] = input[x * (iW-1)/(oW-1)]`.

    This is probably the most natural implementation, however, it has the unwanted side effect where results of scaling something 2x depends on the size. E.g., (assume linear interpolation),
    ```
    input = [0, 1]
    output = [ input[0/3], input[1/3], input[2/3], input[3/3] ]
           = [0, 1/3, 2/3, 1]
   
    input = [0, 1, 0]
    output = [ input[0/5], input[2/5], input[4/5], input[6/5], input[8/5], input[10/5] ]
            = [0, 0.4, 0.8, 0.8, 0.4, 1]
    ```
    As you can see, even though the first two elements are the same, scaling 2x gives different results (in the first couple elements of the output vector). This could be particularly undesirable for image models that are supposed to run on arbitrary spatial sizes.

2. Map `0: out => 0: in`, and `oW: out => iW: in`, i.e., align the corners of the corner pixels, and evaluate at the mapped pixel **centers**, i.e., `output[x] = input[(x+0.5)*iW/oW-0.5]`. 

    ```
    input = [0, 1]
    output = [ input[0.5/2-0.5], input[1.5/2-0.5], input[2.5/2-0.5], input[3.5/2-0.5] ]
           = [0, 1/4, 3/4, 1]
   
    input = [0, 1, 0]
    output = [ input[0.5/2-0.5], input[1.5/2-0.5], input[2.5/2-0.5], input[3.5/2-0.5], input[4.5/2-0.5], input[5.5/2-0.5] ]
            = [0, 1/4, 3/4, 3/4, 1/4, 0]
    ```

**Comment by albanD** [comment]
Thanks for the explanation ! It is indeed non-trivial!

@ZisIsNotZis 
So in the case `align_corners=False` and there is a single output, we expect to always read `input[0/foo]` and so we only get `input[0]` which is an exact entry and thus the current behavior.

If you set `align_corners=True` on the other hand, you will get `input[0.5*iW - 0.5]` which will land your in the middle of the input and will give the behavior you expect.

Is that right?

**Comment by ZisIsNotZis** [comment]
I tried changing align_corners, but it didn't really correct anything

```python
a = torch.rand(1, 1, 2, 2, requires_grad=True)

b = a.mean()
b.backward()
print(b, a.grad, sep='\n')
#tensor(0.3611, grad_fn=<MeanBackward0>)
#tensor([[[[0.2500, 0.2500],
#          [0.2500, 0.2500]]]])

a.grad = None
b = F.avg_pool2d(a, 2)
b.backward()
print(b, a.grad, sep='\n')
#tensor([[[[0.3611]]]], grad_fn=<AvgPool2DBackward>)
#tensor([[[[0.2500, 0.2500],
#          [0.2500, 0.2500]]]])

a.grad = None
b = F.interpolate(a, 1, None, 'bilinear', align_corners=True)
b.backward()
print(b, a.grad, sep='\n')
#tensor([[[[0.3731]]]], grad_fn=<UpsampleBilinear2DBackward>)
#tensor([[[[1., 0.],
#          [0., 0.]]]])

a.grad = None
b = F.interpolate(a, 1, None, 'bilinear', align_corners=False)
b.backward()
print(b, a.grad, sep='\n')
#tensor([[[[0.3731]]]], grad_fn=<UpsampleBilinear2DBackward>)
#tensor([[[[1., 0.],
#          [0., 0.]]]])

a.grad = None
b = F.interpolate(a, 1, None, 'bicubic', align_corners=True)
b.backward()
print(b, a.grad, sep='\n')
#tensor([[[[0.3731]]]], grad_fn=<UpsampleBicubic2DBackward>)
#tensor([[[[1., 0.],
#          [0., 0.]]]])

a.grad = None
b = F.interpolate(a, 1, None, 'bicubic', align_corners=False)
b.backward()
print(b, a.grad, sep='\n')
#tensor([[[[0.4217]]]], grad_fn=<UpsampleBicubic2DBackward>)
#tensor([[[[ 1.1963, -0.1025],
#          [-0.1025,  0.0088]]]])
```

**Comment by albanD** [comment]
Well the fact that `align_corners` is just not used in the kernel function for bilinear [here](https://github.com/pytorch/pytorch/blob/420b37f3c67950ed93cd8aa7a12e673fcfc5567b/aten/src/ATen/native/UpSampleBilinear2d.cpp#L13-L43) looks quite fishy ...

I guess you only get the old behavior for bilinear and the right thing for bicubic right?

**Comment by ZisIsNotZis** [comment]
If "right" means taking information from more or less the center rather than the upper left corner, `bicubic` is also incorrect in both case:

```python
In [4]: a = torch.rand(1, 1, 5, 5, requires_grad=True) 
   ...: a                                                                                                                                                                                                                             
Out[4]: 
tensor([[[[0.1337, 0.7680, 0.9202, 0.6037, 0.6305],
          [0.1870, 0.1369, 0.7258, 0.7428, 0.7928],
          [0.5519, 0.6939, 0.5155, 0.4238, 0.3795],
          [0.0424, 0.8460, 0.0879, 0.5017, 0.2294],
          [0.7575, 0.6702, 0.5161, 0.2578, 0.8698]]]], requires_grad=True)

In [5]: b = F.interpolate(a, 1, None, 'bicubic', align_corners=False) 
   ...: b.backward() 
   ...: a.grad, b                                                                                                                                                                                                                     
Out[5]: 
(tensor([[[[ 1.1963, -0.1025,  0.0000,  0.0000,  0.0000],
           [-0.1025,  0.0088,  0.0000,  0.0000,  0.0000],
           [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
           [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
           [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000]]]]),
 tensor([[[[0.0632]]]], grad_fn=<UpsampleBicubic2DBackward>))

In [6]: a.grad = None 
   ...: b = F.interpolate(a, 1, None, 'bicubic', align_corners=True) 
   ...: b.backward() 
   ...: a.grad, b                                                                                                                                                                                                                     
Out[6]: 
(tensor([[[[1., 0., 0., 0., 0.],
           [0., 0., 0., 0., 0.],
           [0., 0., 0., 0., 0.],
           [0., 0., 0., 0., 0.],
           [0., 0., 0., 0., 0.]]]]),
 tensor([[[[0.1337]]]], grad_fn=<UpsampleBicubic2DBackward>))
```

This highlights the "left-corner" effet

**Comment by albanD** [comment]
Ok,
So the problems are:
- 'bilinear' does not respect the 'align_corners=False' flag.
- both 'bilinear' and 'bicubic' do not follow the formula from @SsnL above: `output[x] = input[(x+0.5)*iW/oW-0.5]` as they return the center of the top left corner, instead of the center of the image. [here](https://github.com/pytorch/pytorch/blob/ec7bb9de1c5f46a9a6fef60c4c80aca126641b8c/aten/src/ATen/native/UpSample.h#L114-L163) are the formulas used in the code.

Is that all the issues you see? I'm trying to get the list of the problems you have to have a well defined task to fix this.

**Comment by ZisIsNotZis** [comment]
In addition, with `oW=1`, the first formula would have `0*iW/0`  which becomes undefined. According to the mapping rule, `0 -> 0`, `(oW-1)=0 -> (iW-1)`, I feel it should also return some symmetric average. Therefore, "centered" rather than "cornered" no matter what `align_corners` is (both symmetric, but probably different)
